### ISSUE TITLE
Python API: Parent ParameterCollection declared locally is destroyed at runtime

### ISSUE THREAD

**Issue by pcyin** [clab/dynet#954]
Hi, after quite a struggle I found Python runtime would destroy parent `ParameterCollection` once it's out of scope, causing all child `ParameterCollection`s unusable.

The story is, I have a Python object of `Parser(args, parent_model)`,  whose constructor looks like:
```
def __init__(self, args, parent_model):
    if parent_model:
        self.model = parent_model.add_subcollection('/parser')
    else:
        _model = dy.ParameterCollection()
        self.model = _model.add_subcollection('/parser')
```

During training, `Parser` is optimized together with other components (with their own `ParameterCollection`s), so I use `parent_model` to store all parameters to optimize, while at test time I just need the `Parser` object, without `parent_model`.

However, the above code will cause segmentation fault when `parent_model=None`. This is because (my hypothesis) `_model` is destroyed by runtime after the `__init__()` call, causing all subcollections to be destroyed. This shouldn't have happened since I assume the subcollection `/parser` will hold a reference to the temporary `_model` object, and in this case `_model` will not be destroyed. However, this happens only in C++ and the Python wrapper  `ParameterCollection` does not keep track of its parent/child parameter collections.

**Comment by pcyin** [comment]
Oh it's related to #930

**Comment by danielhers** [comment]
It's the same problem. I don't think it's just subcollections though, but also parameters. We might want to get the parameters from a trained models to use them, and then we're not interested in the `ParameterCollection` itself any more.

**Comment by neubig** [comment]
This is an consequence of how we implemented DyNet parameter collections (all parameters are stored in the ParameterCollection, so if it goes out of scope the parameters are destroyed, even if there are still remaining references to them). There is a simple solution: never let your top-level ParameterCollection go out of scope. But if you're not aware of this, this could lead to really difficult-to-identify errors. This is a problem in C++ as well.

I think for Python, @pcyin's proposed solution should work. It would add a little bit of overhead for each parameter, but I think it's probably worth it as a temporary fix for this bug.

**Comment by danielhers** [comment]
Fixed by #986, see https://github.com/clab/dynet/blob/master/tests/python/test.py#L237
